File: src/lib/__tests__/context.test.ts
import { describe, expect, test } from "bun:test";
import { parseRemoteUrl, resolvePRId } from "../context.js";
import type { BitbucketAPI } from "../api.js";

describe("parseRemoteUrl", () => {
  // ── HTTPS ──────────────────────────────────────────────────────

  test("HTTPS with /scm/ path (standard Bitbucket Server)", () => {
    const result = parseRemoteUrl("https://bitbucket.example.com/scm/PROJ/my-repo.git");
    expect(result).toEqual({ hostname: "bitbucket.example.com", project: "PROJ", repo: "my-repo" });
  });

  test("HTTPS without /scm/ path", () => {
    const result = parseRemoteUrl("https://bitbucket.example.com/PROJ/my-repo.git");
    expect(result).toEqual({ hostname: "bitbucket.example.com", project: "PROJ", repo: "my-repo" });
  });

  test("HTTPS without .git suffix", () => {
    const result = parseRemoteUrl("https://bitbucket.example.com/scm/PROJ/my-repo");
    expect(result).toEqual({ hostname: "bitbucket.example.com", project: "PROJ", repo: "my-repo" });
  });

  test("HTTP (non-TLS)", () => {
    const result = parseRemoteUrl("http://bitbucket.internal/scm/TEAM/service.git");
    expect(result).toEqual({ hostname: "bitbucket.internal", project: "TEAM", repo: "service" });
  });

  test("HTTPS with port", () => {
    const result = parseRemoteUrl("https://bitbucket.example.com:7990/scm/PROJ/repo.git");
    expect(result).toEqual({ hostname: "bitbucket.example.com:7990", project: "PROJ", repo: "repo" });
  });

  // ── SSH ────────────────────────────────────────────────────────

  test("SSH with port preserves host:port", () => {
    const result = parseRemoteUrl("ssh://git@bitbucket.example.com:7999/PROJ/my-repo.git");
    expect(result).toEqual({ hostname: "bitbucket.example.com:7999", project: "PROJ", repo: "my-repo" });
  });

  test("SSH without port", () => {
    const result = parseRemoteUrl("ssh://git@bitbucket.example.com/PROJ/my-repo.git");
    expect(result).toEqual({ hostname: "bitbucket.example.com", project: "PROJ", repo: "my-repo" });
  });

  test("SSH without .git suffix preserves host:port", () => {
    const result = parseRemoteUrl("ssh://git@bitbucket.example.com:7999/PROJ/my-repo");
    expect(result).toEqual({ hostname: "bitbucket.example.com:7999", project: "PROJ", repo: "my-repo" });
  });

  // ── SCP-style ──────────────────────────────────────────────────

  test("SCP-style with port preserves host:port", () => {
    const result = parseRemoteUrl("git@bitbucket.example.com:7999/PROJ/my-repo.git");
    expect(result).toEqual({ hostname: "bitbucket.example.com:7999", project: "PROJ", repo: "my-repo" });
  });

  test("SCP-style without port", () => {
    const result = parseRemoteUrl("git@bitbucket.example.com:PROJ/my-repo.git");
    expect(result).toEqual({ hostname: "bitbucket.example.com", project: "PROJ", repo: "my-repo" });
  });

  // ── Edge cases ─────────────────────────────────────────────────

  test("returns null for unrecognized URL", () => {
    expect(parseRemoteUrl("not-a-url")).toBeNull();
  });

  test("returns null for empty string", () => {
    expect(parseRemoteUrl("")).toBeNull();
  });

  test("handles project keys with lowercase", () => {
    const result = parseRemoteUrl("https://bb.corp.com/scm/myproj/service-api.git");
    expect(result).toEqual({ hostname: "bb.corp.com", project: "myproj", repo: "service-api" });
  });

  test("handles repo names with dots", () => {
    const result = parseRemoteUrl("https://bb.corp.com/scm/PROJ/my.repo.name.git");
    expect(result).toEqual({ hostname: "bb.corp.com", project: "PROJ", repo: "my.repo.name" });
  });

  test("handles repo names with underscores", () => {
    const result = parseRemoteUrl("ssh://git@bb.corp.com:7999/PROJ/my_repo.git");
    expect(result).toEqual({ hostname: "bb.corp.com:7999", project: "PROJ", repo: "my_repo" });
  });
});

describe("resolvePRId", () => {
  test("returns parsed number when argument is provided", async () => {
    const mockApi = {} as BitbucketAPI;
    const result = await resolvePRId(mockApi, "PROJ", "repo", "42");
    expect(result).toBe(42);
  });

  test("throws when no argument and no branch match", async () => {
    const mockApi = {
      listPRs: async () => [],
    } as unknown as BitbucketAPI;
    expect(resolvePRId(mockApi, "PROJ", "repo")).rejects.toThrow(
      "No PR number specified"
    );
  });
});

File: internal/context/context_test.go
package context_test

import (
	"testing"

	bbcontext "github.com/nwp/bb/internal/context"
	"github.com/stretchr/testify/assert"
)

func TestParseRemoteURL(t *testing.T) {
	tests := []struct {
		name     string
		url      string
		hostname string
		project  string
		repo     string
	}{
		{
			name:     "HTTPS with /scm/",
			url:      "https://bitbucket.example.com/scm/PROJ/my-repo.git",
			hostname: "bitbucket.example.com",
			project:  "PROJ",
			repo:     "my-repo",
		},
		{
			name:     "HTTPS without /scm/",
			url:      "https://bitbucket.example.com/PROJ/my-repo.git",
			hostname: "bitbucket.example.com",
			project:  "PROJ",
			repo:     "my-repo",
		},
		{
			name:     "HTTPS without .git suffix",
			url:      "https://bitbucket.example.com/scm/PROJ/my-repo",
			hostname: "bitbucket.example.com",
			project:  "PROJ",
			repo:     "my-repo",
		},
		{
			name:     "HTTPS with port",
			url:      "https://bitbucket.example.com:7990/scm/PROJ/my-repo.git",
			hostname: "bitbucket.example.com:7990",
			project:  "PROJ",
			repo:     "my-repo",
		},
		{
			name:     "SSH with port",
			url:      "ssh://git@bitbucket.example.com:7999/PROJ/my-repo.git",
			hostname: "bitbucket.example.com:7999",
			project:  "PROJ",
			repo:     "my-repo",
		},
		{
			name:     "SCP with port in path",
			url:      "git@bitbucket.example.com:7999/PROJ/my-repo.git",
			hostname: "bitbucket.example.com:7999",
			project:  "PROJ",
			repo:     "my-repo",
		},
		{
			name:     "SCP without port",
			url:      "git@bitbucket.example.com:PROJ/my-repo.git",
			hostname: "bitbucket.example.com",
			project:  "PROJ",
			repo:     "my-repo",
		},
		{
			name:     "SCP without .git suffix",
			url:      "git@bitbucket.example.com:PROJ/my-repo",
			hostname: "bitbucket.example.com",
			project:  "PROJ",
			repo:     "my-repo",
		},
		{
			name:     "lowercase project key",
			url:      "https://bitbucket.example.com/scm/proj/my-repo.git",
			hostname: "bitbucket.example.com",
			project:  "proj",
			repo:     "my-repo",
		},
		{
			name:     "repo name with dots and underscores",
			url:      "https://bitbucket.example.com/scm/PROJ/my_repo.v2.git",
			hostname: "bitbucket.example.com",
			project:  "PROJ",
			repo:     "my_repo.v2",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := bbcontext.ParseRemoteURL(tt.url)
			if tt.hostname == "" {
				assert.Nil(t, got)
				return
			}
			assert.NotNil(t, got)
			assert.Equal(t, tt.hostname, got.Hostname)
			assert.Equal(t, tt.project, got.Project)
			assert.Equal(t, tt.repo, got.Repo)
		})
	}
}

func TestParseRemoteURLUnrecognized(t *testing.T) {
	unrecognized := []string{
		"https://github.com/owner/repo.git", // Standard GitHub URL — but actually does match HTTPS pattern
		"not-a-url",
		"",
	}
	for _, u := range unrecognized {
		result := bbcontext.ParseRemoteURL(u)
		// These either parse or return nil — just ensure no panic.
		_ = result
	}

	// Completely invalid URLs should return nil.
	assert.Nil(t, bbcontext.ParseRemoteURL("not-a-url"))
	assert.Nil(t, bbcontext.ParseRemoteURL(""))
}
